src/algebra/regular.lean

```diff
@@ -0,0 +1,103 @@
+/-
+Copyright (c) 2021 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import algebra.group
+
+/-!
+# Regular elements
+
+We introduce left-regular, right-regular and regular elements.
+
+By definition, a regular element in a commutative ring is a non-zero divisor.
+Lemma `is_regular_of_integral_domain` shows that every non-zero element of an integral domain
+is regular.
+
+The final goal is to develop part of the API to prove, eventually, results about non-zero-divisors.
+-/
+variables {R : Type*}
+
+/-- A left-regular element is an element `c` such that multiplication on the left by `c`
+is injective on the left. -/
+def is_left_regular [has_mul R] (c : R) := function.injective ((*) c)
+
+/-- A right-regular element is an element `c` such that multiplication on the right by `c`
+is injective on the right. -/
+def is_right_regular [has_mul R] (c : R) := function.injective (* c)
+
+/-- A regular element is an element `c` such that multiplication by `c` both on the left and
+on the right is injective. -/
+structure is_regular [has_mul R] (c : R) : Prop :=
+(left : is_left_regular c)
+(right : is_right_regular c)
+
+namespace is_regular
+
+/-- If multiplication is associative, then the product of left-regular elements is
+left-regular. -/
+lemma left_mul [has_mul R] (mul_assoc : ∀ x y z : R, x * y * z = x * (y * z)) {a b : R}
```

Comment:
`has_mul M` + `mul_assoc` is `semigroup R`

Reply:
I was convinced that I tried using `semigroup` and got an error.  Anyway, thanks and it works now!